source/TimeSeries.mc

```diff
@@ -0,0 +1,436 @@
+import Toybox.Lang;
+import Toybox.Test;
+using Toybox.System;
+
+typedef StatePoint as Array<Number or Float>; // [ts, percent]
+typedef StatePoints as Array<StatePoint>;
+
+
+(:background)
+class TimeSeriesOpts {
+    public static const SIZE = 8; // bytes used to store TimeSeriesOpts
+    private const START_SIZE = 4; // bytes for storing start timestamp
+    private const OFFSET_BITS = 16; // bytes for storing offset
+    private const OFFSET_MASK = 1 << OFFSET_BITS - 1; // mask for offset
+    private var mStart as Number;  // iterator start timestamp
+    private var mSize as Number; // unpacked size
+    private var mOffset as Number; // zero element offset
+
+    (:debug)
+    public static function FromBytes(b as ByteArray, offset as Number) as TimeSeriesOpts {
+        var opts = new TimeSeriesOpts(0, 0, 0);
+        opts.load(b, offset);
+        return opts;
+    }
+
+    public function initialize(start as Number, size as Number, offset as Number) as Void {
+        mStart = start;
+        mSize = size;
+        mOffset = offset;
+    }
+
+    public function save(b as ByteArray, offset as Number) as Void {
+        b = encodeNumber(b, mStart, offset);
+        offset += START_SIZE;
+        var n = mSize << OFFSET_BITS + mOffset;
+        b = encodeNumber(b, n, offset);
+    }  
+    
+    public function load(b as ByteArray, offset as Number) as Void {
+        mStart = decodeNumber(b, offset);
+        offset += START_SIZE;
+        var n = decodeNumber(b, offset);
+        mSize = n >> OFFSET_BITS;
+        mOffset = n & OFFSET_MASK;
+    }
+
+    function getStart() as Number {
+        return mStart;
+    }
+
+    function getSize() as Number {
+        return mSize;
+    }
+
+    function getOffset() as Number {
+        return mOffset;
+    }
+}
+
+(:background)
+class TimeSeries {
+
+    // static const MAX_TS = 1 << 22;
+    // const MAX_VALUE = 1 << 10;  // right 10 bits
+    // const INT32_MASK = 1l << 32;
+    // const INT16_MASK = 1l << 16;
+    // const LOW_MASK = INT32_MASK - 1;
+    // const HIGH_MASK = LOW_MASK << 32;
+
+    /*
+    int32 format:
+    <ts 22 bit as offset in seconds><value 10 bit as battery ppm>
+
+    * 48 days vector duration (2^22 seconds)
+    * ts is stored as offset from start
+    * value accuracy is 0.1%
+
+    last array value contains: (start ts) << 32 | size << 16 | offset
+
+    */
+    private var mPoints as ByteArray;  // packed uint32 points + int64 opts
+    private var mSize as Number; // unpacked size
+    private var mStart as Number;  // iterator start timestamp
+    private var mOffset as Number; // zero element offset
```

Comment:
replace with TimeSeriesOpts?

Reply:
no, because these variables change in add/set etc... methods, and if move them to opts, nothing meaningful is left in TimeSeries class,